src/txrebroadcast.cpp

```diff
@@ -0,0 +1,242 @@
+            // We have already rebroadcast this transaction the maximum number
+            // of times permitted, so skip rebroadcasting.
+            continue;
+        } else if (entry_it->m_last_attempt > start_time - MIN_REATTEMPT_INTERVAL) {
+            // We already rebroadcasted this in the past 4 hours. Even if we
+            // added it to the set, it would probably not get INVed to most
+            // peers due to filterInventoryKnown.
+            continue;
+        } else {
+            // We have rebroadcasted this transaction before, but will try
+            // again now. Record the attempt.
+            auto UpdateRebroadcastEntry = [start_time](RebroadcastEntry& rebroadcast_entry) {
+                rebroadcast_entry.m_last_attempt = start_time;
+                ++rebroadcast_entry.m_count;
+            };
+
+            m_attempt_tracker->modify(entry_it, UpdateRebroadcastEntry);
+        }
+
+        // Add to set of rebroadcast candidates
+        rebroadcast_txs.push_back(TxIds(txid, wtxid));
+    }
+
+    TrimMaxRebroadcast();
+
+    auto delta1 = after_cnb_time - start_time;
+    auto delta2 = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "GetRebroadcastTransactions(): %d us total, %d us spent in CreateNewBlock.\n", delta2.count(), delta1.count());
+    LogPrint(BCLog::NET, "Queued %d transactions for attempted rebroadcast, filtered from %d candidates with cached fee rate of %s.\n", rebroadcast_txs.size(), block_template->block.vtx.size() - 1, m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB));
+
+    for (TxIds ids : rebroadcast_txs) {
+        LogPrint(BCLog::NET, "Attempting to rebroadcast txid: %s, wtxid: %s\n", ids.m_txid.ToString(), ids.m_wtxid.ToString());
+    }
+
+    return rebroadcast_txs;
+};
+
+void TxRebroadcastHandler::CacheMinRebroadcastFee()
+{
+    if (m_chainman.ActiveChainstate().IsInitialBlockDownload()) return;
+
+    LOCK(m_rebroadcast_mutex);
+
+    // Update stamp of chain tip on cache run
+    m_tip_at_cache_time = m_chainman.ActiveTip();
+
+    // Store the existing fee rate
+    m_previous_cached_fee_rate = m_cached_fee_rate;
+
+    // Calculate a new cached fee rate
+    auto start_time = GetTime<std::chrono::microseconds>();
+    m_cached_fee_rate = BlockAssembler(m_chainman.ActiveChainstate(), m_mempool, m_chainparams).MinTxFeeRate();
+    auto delta_time = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "Caching minimum fee for rebroadcast to %s, took %d us to calculate.\n", m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB), delta_time.count());
+};
+
+void TxRebroadcastHandler::RemoveFromAttemptTracker(const CTransactionRef& tx) {
+    LOCK(m_rebroadcast_mutex);
+    const auto it = m_attempt_tracker->find(tx->GetWitnessHash());
+    if (it == m_attempt_tracker->end()) return;
+    m_attempt_tracker->erase(it);
+}
+
+void TxRebroadcastHandler::TrimMaxRebroadcast()
```

Comment:
It seems to me that `m_attempt_tracker` has two distinct purposes:

1. prevent us from rebroadcasting transactions too frequently (see `entry_it->m_last_attempt > start_time - MIN_REATTEMPT_INTERVAL`)
2. prevent us from rebroadcasting transactions too many times (see `else if (entry_it->m_count >= MAX_REBROADCAST_COUNT)`)

I think the expiry strategies conflict for these two purposes. Imagine that we (unupgraded) have 550 txs that are no longer standard for the rest of the network

- for don't-rebroadcast-too-frequently, we want to expire least-recently-rebroadcast first, since if we expire more recently rebroadcast transactions, then we'll try to rebroadcast them again the next time we receive a block, and we'll end up rebroadcasting a very small set of txs very frequently.
- for don't-rebroadcast-too-many-times, if we expire the least-recently-rebroadcast first, then the next time we calculate a rebroadcast set, we'll include them again, re-add them to attempt_tracker, and expire the next oldest 50. We'll continue cycling all the now-nonstandard transactions through the attempt tracker over and over again.

I think it might be a good idea to separate these two purposes by adding a (rolling?) bloom filter for (2). Once a transaction has been rebroadcast `MIN_REATTEMPT_INTERVAL` times (or it's removed from the attempt tracker because the tracker is full) we toss it in the bloom filter and then never rebroadcast it again (at least until restart).